### DISCUSSION TITLE
Is nn-Unet perform well on smaller datasets? What would be the minimum number of data needed?

### DISCUSSION THREAD

**Discussion by iromimp** [MIC-DKFZ/nnUNet#1166] [Q&A]
Hello,

I do have a really small dataset where I have around 8 patients which I have to work on. I really do want to work on nn-unet and just wondering whether running an nn-Unet model on that dataset is worth getting an idea.  What would be the minimal number of data needed to get a good performance in nn-Unet?  I really appreciate it if you have any guidance or suggestions to come up with a workaround. 

Thank You

**Comment by FabianIsensee** [comment]
The amount of training data needed strongly depends on the segmentation problem. The more difficult it is and the more diverse the appearance of the images/target structures the more training cases you need. This is true for all segmentation methods, not just nnU-Net.
The only way to find out if it's enough is to run it and look at the results
Best,
Fabian

**Reply by iromimp** [reply]
Hi Fabian,

Thank you so much for the clarification. I have run the model with the data and I got low results on the test data where dice similarity is 0.39. I have used both PET and CT data to run the model. My target was to identify bone metastases. I have input CT scans without masking the bone region and the PET scans as the input data. I used 6 patients as training data and 2 patients as test data.
Guess I need to find more data on bone metastases which is not available online.

Thank You
Best Regards
Iromi

**Reply by Joeycho** [reply]
How big image size do you use for one patient? Is it 3D image or 2D image?

I think identifying bone metastatses seems way harder task than mine(bone and teeth), but I already did with 12 training patients and 4 test patients. The result was very satisfied. That caused a bit problem, but I am interested in even reducing more..

I think that patch-approach and generating each batch randomly enable higher performance even on smaller datasets with some conditions. In my case, I have 3D MR and CT images and they are quite big enough and quality is also good. The default number of batches 250 was enough for me within 30 epochs.

Maybe you can try even increasing the default number of batches (250 -> 500) or increase the maximum number of epochs.  Or adjust data augmentation for your data (enable or disable.. or add more variability in spatial perspective, rather than intensity ones, I guess your task might be sensitive on intensities...).

**Reply by iromimp** [reply]
Hi,

Thank you so much for your reply. The image size of the patients varies as follows.
-  Training data: 5 patients
       3 patients: 512,512,263
       2 patients: 512,512,287

- Test data: 2 patients
      512,512,263
      512,512,311
All the images are 3D.
I will incorporate your suggestions ad try. So far, I didn't make any changes to the default conditions in the framework. 

Thank You
Best Regards
Iromi

**Reply by Joeycho** [reply]
Hi,

Are they original sizes of the images? How about preprocessed images? cropped ones by nnUNet preprocessing?

Plus, it might be helpful if you can share part of debug.json file as well, especially final_ patch size. And probably, the number of epochs and the number of batches for your past training (0.39) one.

**Reply by iromimp** [reply]
Hi,

Yes, those are original image sizes. The preprocessed and cropped ones are also the same, but in (z,y,x) order (ex: 311,512,512). Herewith I attach the debug.json file for reference.
Number of epochs: 1000
Number of batches: 250 batches per epoch
Thank you very much for helping me to resolve this.

[debug.txt](https://github.com/MIC-DKFZ/nnUNet/files/9650573/debug.txt)

Thank you
Best Regards
Iromi

**Reply by Joeycho** [reply]
Hi,

I'm not sure it might help you... but  you may try to train 5 fold training, not 'all'. Since you have 5 training samples, you can still do 5 fold cross-validation training. Now, with 'all', 5 samples are exposed for training and validation.

How much does it it take for 1 epoch? 1000 seems really enough, but I wonder the duration for 1 epoch. Did you finish it upto 1000? And how does loss and metric graph look like (progress.png)?

**Reply by iromimp** [reply]
Hi, 
I did do the 5 fold training as well. But, the results are not good.  1 epoch took an average of 120 seconds to finish and training finish up to 1000. The progress graph looks fine to me though. 
Here is the progress.png for fold= all
![progress fold all](https://user-images.githubusercontent.com/79505848/192644086-b690a887-5b6b-4a5c-bca9-c40e8aa58f1b.png)
The progress of fold 3 is as follows. The progress of all the folds is similar to the below graph.
![progress fold 3](https://user-images.githubusercontent.com/79505848/192645069-7226986c-395f-40ca-928d-3e9202b2f5ae.png)

Thank You
Best Regards
Iromi

**Reply by Joeycho** [reply]
Hmm.. 

fold 3, (and every fold), it looks weird graph. So.. maybe it overfit to training set and did not improve with validation set.. Even getting worse..

Can you share 'splits_final.pkl' in processed folder? I wonder how folding is done in your training..

**Reply by iromimp** [reply]
Hi,

Thank you for the reply and yes it is. Please find the below zip file of the splits_final.pkl.
[splits_final.zip](https://github.com/MIC-DKFZ/nnUNet/files/9663862/splits_final.zip)
Thank You
Best Regards
Iromi

**Reply by Joeycho** [reply]
It looks fine.. but 'ANON67365', 'GH_SABR_03', is there any difference between ANON... and GH..?

**Reply by iromimp** [reply]
Hi,
It's just a name. No difference at all. 2 of the patients has more than 30 mets and rest vary in between 2 to 6. 

Thank you
Best Regards 
Iromi

**Reply by Joeycho** [reply]
Hmm... okay. It's very weird, since evaluation metric under 0.01.. means... it probably failed to predict anything with validation set..

With 'all', at least training is working and if just do validation with the training set, it seems fine. And did you get 0.39 with trained model from 'all'? right?

**Reply by iromimp** [reply]
Hi,

Yes. 0.39 is for "all". I couldn't figure out any reason for poor performance.

Thank You
Best Regards
Iromi

**Reply by Joeycho** [reply]
OrderedDict([('train', array(['ANON50135', 'ANON51274', 'ANON67365', 'GH_SABR_03'], dtype='<U10')), ('val', array(['ANON82986'], dtype='<U10'))])

Hi, so the above is matched with fold 3, train and validation list. Maybe you can try the exact same training set like the above, but train with 'all', and test on 'ANON82986'? The result should be similar as described in the 2nd progress plot.

If the result is actually similar with 1st one, I think maybe we can try to expose some training set to validation set, too and do 5 folding..

**Reply by iromimp** [reply]
Thank you for the suggestion. I will do the same and let you know the results.

Thanks again

**Reply by iromimp** [reply]
Hi, 

Sorry for the late reply. I have trained a model with fold 3 but "all".  But I got a dice score of 0. But the progress plot is better as below.
![progress_with all](https://user-images.githubusercontent.com/79505848/196819557-24b29dc3-27bc-473f-8a5a-77f8ac730d77.png)

 Then I tried the training with data augmentation ([nnUNetTrainerV2_DA5.py](https://github.com/MIC-DKFZ/nnUNet/blob/master/nnunet/training/network_training/nnUNet_variants/data_augmentation/nnUNetTrainerV2_DA5.py)). Yet I am getting a dice score as 0. 
Could you please let me know what tools you used to convert the Dicom data to nifti?
I would like to know if is there any possibility for you to share the dataset you used for me to go through the experiment.

Thank You
Best Regards
Iromi

**Reply by Joeycho** [reply]
Hi Iromi,

Unfortunately, the dataset is not owned by me, and as you might imagine.. it is somehow.. confidential. I cannot share it to you...

From DICOM to nifti, I've used pydicom libarary to read dicom folder, and save it as nifti with sitk.

import SimpleITK as sitk
from pydicom.filereader import dcmread

sitk.ReadImage with option "GDCMImageIO", or sitk.ImageSeriesReader().

I think there are many other tools too. 

Regarding the above graph, it seems fine!. Metric goes upto 0.9, that means training is working, although you use 'all' setting, which means exposing validation set in training. Maybe then.. Could you try another fold(4), and predict one of data, which was training data in fold 3, but not in this new fold (4).

Or another check point is just visualize those data(one data in test (dice 0), and one data from training set). and see whether they are similar or not.. If they look quite different, then it makes sense.. Source data and target data comes from different distribution which makes the task difficult..
